### ISSUE TITLE
IPv6 DHT implementation does not follow BEP32 or 45 socket binding rules.

### ISSUE THREAD

**Issue by the8472** [arvidn/libtorrent#1266]
I'm observing the following misbehavior coming from ipv6 libtorrent nodes

_[Error] mtid matched, socket address did not, ignoring message, request: /2607:fb90:40a0:2ce5:0:27:bf48:b201:6881 -> response: /2603:3020:2d05:1d00:d819:d2e4:e4cb:ccd7:6881 v:0x4C540102/LT��_

I.e. the a response was sent from a different socket address than the request was sent to, which indicates that it was bound to `::` instead of a specific address.

http://bittorrent.org/beps/bep_0032.html#source-address-selection
http://bittorrent.org/beps/bep_0045.html#mandatory

**Comment by the8472** [comment]
Also note that just sending a response from the socket address the request was received on, while an improvement, would not be sufficient since it would still expose the node ID on multiple addresses.

**Comment by aldenml** [comment]
Just to confirm that in fact, the current (mis)behavior is the coded one:
```
LOG - log_alert - applying settings pack, init=true, reopen_listen_port=false
LOG - log_alert - update listen interfaces: 0.0.0.0:6881,[::]:6881
LOG - log_alert - parsed listen interfaces count: 2, ifaces: 0.0.0.0:6881,[::]:6881
LOG - log_alert - reopen listen sockets
LOG - log_alert - no currently open sockets to close
LOG - log_alert - attempting to open listen socket to: 0.0.0.0:6881 on device:  flags: 0
LOG - log_alert -  listening on: 0.0.0.0 TCP port: 6881 UDP port: 6881
LOG - log_alert - attempting to open listen socket to: [::]:6881 on device:  flags: 0
LOG - log_alert -  listening on: :: TCP port: 6881 UDP port: 6881
LISTEN_SUCCEEDED - listen_succeeded_alert - successfully listening on [TCP] 0.0.0.0:6881
LISTEN_SUCCEEDED - listen_succeeded_alert - successfully listening on [UDP] 0.0.0.0:6881
LISTEN_SUCCEEDED - listen_succeeded_alert - successfully listening on [TCP] [::]:6881
LISTEN_SUCCEEDED - listen_succeeded_alert - successfully listening on [UDP] [::]:6881
```

**Comment by ssiloti** [comment]
Multi-home support is still a work-in-progress. Arvid was working on it at one point but I don't think it's been touched in a while. IIRC there's some significant refactoring that needs to be done to the way libtorrent handles UDP in order to properly support multiple interfaces.

**Comment by the8472** [comment]
well, multi-home support is optional, just picking one also is a valid approach.

**Comment by arvidn** [comment]
if only there was better, platform independent, support for enumerating interfaces and IPs associated with them. I have this functionality, but it seems like every time I touch it I break something...

**Comment by the8472** [comment]
What about the connect+getsocketname approach? transmission's and my DHT use that for single-address operation.

**Comment by ssiloti** [comment]
I think we should bite the bullet and implement proper multi-home support, at least for IPv6. One thing we're going to need before we ban binding to `::` is the ability to subscribe to IP change notifications from the OS so that we can reliably maintain sockets for all available addresses. Of course this is also highly platform specific. I'm working on it for Linux and Windows, but I don't have access to a Mac so OSX support will have to be done by someone else.

**Comment by the8472** [comment]
see also #1081

Basically, there are 3 valid implementation tiers:

1. single-address. pick a unicast address and bind
2. separate multi-address. just run N separate DHT instances that don't have to talk much to each other for each address. that's basically an extension of the v4/v6 split which already requires multiple instances to run
3. shared multi-address. this has some shared state for a whole address family but also separate state per socket. it allows for some optimizations but increases complexity

**Comment by ssiloti** [comment]
A tier 2 implementation is now in master.
The bulk of the changes are in cdd50be859a470433e7dee6f90fc29c42912966. fd50630020cbba5ee428aa15cd325e301b1c4658 adds the final change to bind to individual IPv6 interfaces by default instead of `::`.

**Comment by the8472** [comment]
I have found some even more peculiar behavior. Maybe this is going to be fixed once people deploy newer master builds, but it seems quite unusual.

```
[2017-09-20T19:34:47.786012Z][Error] mtid matched, socket address did not, ignoring message, request: /2406:5a00:7006:3600:ae5f:3eff:fe67:ce0f:42725 -> response: /fc01:0:0:0:9c44:2a2:2047:b9b:42725 v:0x4C540102/LT��
[2017-09-20T19:34:47.786201Z][Error] mtid matched, socket address did not, ignoring message, request: /2406:5a00:7006:3600:ae5f:3eff:fe67:ce0f:42725 -> response: /2406:5a00:7006:3600:9c44:2a2:2047:b9b:42725 v:0x4C540102/LT��
[2017-09-20T19:34:47.788125Z][Error] mtid matched, socket address did not, ignoring message, request: /2406:5a00:7006:3600:ae5f:3eff:fe67:ce0f:42725 -> response: /fc01:0:0:0:ae5f:3eff:fe67:ce0f:42725 v:0x4C540102/LT��
```

Note that here libtorrent nodes are responding from [rfc4193](https://tools.ietf.org/html/rfc4193) unique local unicast addresses. AIUI this *should not happen* even with multihoming since automatic source address selection should pick something other than a ULA.

**Comment by aldenml** [comment]
I think this commit should address the issue (and it's relatively recent): https://github.com/arvidn/libtorrent/commit/d4986f878f7c67cc8ede7748a241a64fbcdbe950

**Comment by ssiloti** [comment]
Even without d4986f8 we shouldn't be responding to requests using an address other than the one it was sent to. I think the problem is that I forgot to add code in `dht::node::incoming` to filter out requests which are not addressed to that node.

**Comment by aldenml** [comment]
I asked about that "lack of filtering" that allows to one DHT parse any incoming DHT packet here (in this case I asked by the protocol): https://github.com/arvidn/libtorrent/issues/2306

> ... or it is just a secondary effect of passing the incoming packet in dht_tracker::incoming_packet to all nodes (without discriminating protocol)? or it's that the intended behavior?

**Comment by aldenml** [comment]
my point is that maybe having the "secondary effects" in the incoming, that I'm not sure are intended or not, is a good think, and putting the filter in the output is the solution.

**Comment by ssiloti** [comment]
This is not related to #2306. Nodes _do_ filter on address family, so IPv6 requests will not generate responses from IPv4 addresses. That code was added as part of the original IPv6 DHT commit which only supported one node per address family. Now that we have multi-home support that code needs to be changed to match on the listen socket instead of just the address family.

**Comment by aldenml** [comment]
right, I just quickly read `dht::node::incoming to filter out requests` and didn't realize you were in fact talking about "responses"

**Comment by ssiloti** [comment]
Nodes sending multiple responses with different IPs should be fixed as of 8648de3706efe0d9ab2504793bd0dd8a9566a1d6.
